source/fe/fe_simplex_p.cc

```diff
@@ -436,6 +438,138 @@ FE_SimplexPoly<dim, spacedim>::get_prolongation_matrix(
+  // special case for 1D.
+  if (dim == 1)
+    Assert(combined_orientation == numbers::default_geometric_orientation,
+           ExcMessage("In 1D, all faces must have the default orientation."));
+  else
+    AssertIndexRange(combined_orientation,
+                     this->reference_cell().n_face_orientations(face));
+
+  // we need to distinguish between DoFs on vertices, lines and in 3d quads.
+  // do so in a sequence of if-else statements
+  if (face_index < this->get_first_face_line_index(face))
+    // DoF is on a vertex
+    {
+      // get the number of the vertex on the face that corresponds to this DoF,
+      // along with the number of the DoF on this vertex
+      const unsigned int face_vertex = face_index / this->n_dofs_per_vertex();
+      const unsigned int dof_index_on_vertex =
+        face_index % this->n_dofs_per_vertex();
+
+      // then get the number of this vertex on the cell and translate
+      // this to a DoF number on the cell
+      return this->reference_cell().face_to_cell_vertices(
+               face, face_vertex, combined_orientation) *
+               this->n_dofs_per_vertex() +
+             dof_index_on_vertex;
+    }
+  else if (face_index < this->get_first_face_quad_index(face))
+    // DoF is on a line
+    {
+      // do the same kind of translation as before. we need to only consider
+      // DoFs on the lines, i.e., ignoring those on the vertices
+      const unsigned int index =
+        face_index - this->get_first_face_line_index(face);
+
+      const unsigned int face_line         = index / this->n_dofs_per_line();
+      const unsigned int dof_index_on_line = index % this->n_dofs_per_line();
+
+      // we now also need to adjust the line index for the case of
+      // face orientation, face flips, etc
+      unsigned int adjusted_dof_index_on_line = 0;
+      switch (dim)
+        {
+          case 1:
+            DEAL_II_ASSERT_UNREACHABLE();
+            break;
+
+          case 2:
+            if (combined_orientation == numbers::default_geometric_orientation)
+              adjusted_dof_index_on_line = dof_index_on_line;
+            else
+              adjusted_dof_index_on_line =
+                this->n_dofs_per_line() - dof_index_on_line - 1;
+            break;
+
+          case 3:
+            // in 3d, things are difficult. someone will have to think
+            // about how this code here should look like, by drawing a bunch
+            // of pictures of how all the faces can look like with the various
+            // flips and rotations.
+            //
+            // that said, we can implement a couple other situations easily:
+            // if the face orientation is
+            // numbers::default_combined_face_orientation then things are
+            // simple. Likewise if the face orientation is
+            // numbers::reverse_line_orientation then we just know we need to
+            // reverse the DoF order on each line.
```

Comment:
This isn't correct in 3D since the face orientation is not a line orientation: i.e., it is only meaningful to split into those two cases when we have lines.

I don't have time to look into it that carefully right now but there are some additional orientation complexities present here which don't occur in the hypercube case. For example: the lines of the faces on a Tetrahedron are not consistently oriented (which we correct for via `ReferenceCell::face_to_cell_line_orientation()`). Any solution here needs to take that into account.